helix-term/src/grammars.rs

```diff
@@ -0,0 +1,299 @@
+
+        // create the grammar dir contains a git directory
+        if !grammar_dir.join(".git").is_dir() {
+            git(&grammar_dir, ["init"])?;
+        }
+
+        // ensure the remote matches the configured remote
+        if get_remote_url(&grammar_dir).map_or(true, |s| s.trim_end() != remote) {
+            set_remote(&grammar_dir, &remote)?;
+        }
+
+        // ensure the revision matches the configured revision
+        if get_revision(&grammar_dir).map_or(true, |s| s.trim_end() != revision) {
+            // Fetch the exact revision from the remote.
+            // Supported by server-side git since v2.5.0 (July 2015),
+            // enabled by default on major git hosts.
+            git(&grammar_dir, ["fetch", REMOTE_NAME, &revision])?;
+            git(&grammar_dir, ["checkout", &revision])?;
+
+            println!(
+                "Grammar '{}' checked out at '{}'.",
+                grammar.grammar_id, revision
+            );
+            Ok(())
+        } else {
+            println!("Grammar '{}' is already up to date.", grammar.grammar_id);
+            Ok(())
+        }
+    } else {
+        println!("Skipping local grammar '{}'", grammar.grammar_id);
+        Ok(())
+    }
+}
+
+// Sets the remote for a repository to the given URL, creating the remote if
+// it does not yet exist.
+fn set_remote(repository: &Path, remote_url: &str) -> Result<String> {
+    git(repository, ["remote", "set-url", REMOTE_NAME, remote_url])
+        .or_else(|_| git(repository, ["remote", "add", REMOTE_NAME, remote_url]))
+}
+
+fn get_remote_url(repository: &Path) -> Option<String> {
+    git(repository, ["remote", "get-url", REMOTE_NAME]).ok()
+}
+
+fn get_revision(repository: &Path) -> Option<String> {
+    git(repository, ["rev-parse", "HEAD"]).ok()
+}
+
+// A wrapper around 'git' commands which returns stdout in success and a
+// helpful error message showing the command, stdout, and stderr in error.
+fn git<I, S>(repository: &Path, args: I) -> Result<String>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<std::ffi::OsStr>,
+{
+    let output = Command::new("git")
+        .args(args)
+        .current_dir(repository)
+        .output()?;
+
+    if output.status.success() {
+        Ok(String::from_utf8_lossy(&output.stdout).into_owned())
+    } else {
+        // TODO: figure out how to display the git command using `args`
+        Err(anyhow!(
+            "Git command failed.\nStdout: {}\nStderr: {}",
+            String::from_utf8_lossy(&output.stdout),
+            String::from_utf8_lossy(&output.stderr),
+        ))
+    }
+}
+
+fn build_grammar(grammar: GrammarConfiguration) -> Result<()> {
+    let grammar_dir = if let GrammarSource::Local { ref path } = grammar.source {
+        PathBuf::from(path)
+    } else {
+        helix_core::runtime_dir()
+            .join("grammars/sources")
+            .join(grammar.grammar_id.clone())
+    };
+
+    if grammar_dir.read_dir().is_err() {
```

Comment:
We're again assuming that any error only means one thing. I am not super familiar with anyhow, but judging from the readme, it looks like you could just do something like `grammar_dir.read_dir().with_context(|| format!("Failed to read directory {:?}; did you use 'hx --fetch-grammars'?", grammar_dir))`